### DISCUSSION TITLE
slow on rtsp streaming

### DISCUSSION THREAD

**Discussion by Atharva-Malode** [ultralytics/ultralytics#10106] [Q&A]
I tried using object detection on a USB cam and the built-in cam in my laptop, and it worked well with zero lags. However, when I tried using it on an RTSP streaming link, it lagged a lot. I tried changing the buffer size, but it didn't work. I also noticed that I was losing a few frames from the stream. **RTSP cam working slow.** What may be the possible reason?

*Note:* I am trying this by connecting Ethernet cables one end to the IP cam and the other end directly to my laptop. I tried this thinking bandwidth may be the issue.

Can anyone suggest how to not lose frames and improve performance?

```python
import cv2
from ultralytics.utils.plotting import Annotator
from ultralytics import YOLO

rtsp_url = "rtsp://sample:sample@123@192.168.20.19:554/cam/realmonitor?channel=1&subtype=1"

model = YOLO("yolov8n.pt")

cap = cv2.VideoCapture(rtsp_url)
# cap = cv2.VideoCapture(0)

if not cap.isOpened():
    print("Error: Could not open camera.")
    exit()

cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)
cap.set(cv2.CAP_PROP_BUFFERSIZE, 1)

while True:
    _, img = cap.read()

    results = model.predict(img)

    for r in results:
        annotator = Annotator(img)

        boxes = r.boxes
        for box in boxes:
            b = box.xyxy[0]
            c = box.cls
            annotator.box_label(b, model.names[int(c)])

    img = annotator.result()
    cv2.imshow('YOLO V8 Detection', img)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

**Comment by pderrenger** [comment]
@Atharva-Malode it sounds like you've done a great job troubleshooting so far! When experiencing lag with RTSP streams, the issue often boils down to network latency or processing power rather than bandwidth, particularly if you're not facing the same issues with USB or built-in cams. 

A few potential solutions to consider:

1. **Stream Compression/Quality:** Check if the RTSP stream is at a very high quality or bitrate. Adjusting the stream's settings to a lower quality or bitrate might reduce the latency without significantly compromising the detection quality.

2. **Processing Power:** Ensure your laptop's hardware can handle the workload. Object detection, especially with high-resolution video streams, can be resource-intensive.

3. **Threading:** Implement threading or asynchronous processing to prevent the main loop from waiting for each frame to be processed. This can help in not losing frames. Python threading or `concurrent.futures` can be useful here. 

For the threading part, here's a quick snippet to give you an idea:

```python
import threading
import cv2
from ultralytics import YOLO

rtsp_url = "rtsp_url"
model = YOLO("yolov8n.pt")

def process_frame():
    while cap.isOpened():
        _, img = cap.read()
        # Prediction and annotations code here

# Start the capturing
cap = cv2.VideoCapture(rtsp_url)

if not cap.isOpened():
    print("Error: Could not open camera.")
else:
    thread = threading.Thread(target=process_frame)
    thread.start()
```

Note: This is a simplified example, and error checking is recommended for production code.

Hope this helps! Let us know how it goes. 🚀

**Comment by Atharva-Malode** [comment]
@pderrenger , I tried your suggestion, but it didn't work for me. It's fine about the delay, but I don't want to lose frames. I tried using the simple OpenCV library to stream the video, and it's working too well without losing a single frame.

my simple open cv code :

```
import cv2

# rtsp_url = "rtsp://admin:admin@123@192.168.20.19:554/cam/realmonitor?channel=1&subtype=0"
rtsp_url = "rtsp://admin:admin@123@192.168.20.19:554/cam/realmonitor?channel=1&subtype=1"

# Open the RTSP stream
cap = cv2.VideoCapture(rtsp_url)

if not cap.isOpened():
    print("Error: Could not open camera.")
    exit()

# Set video capture properties
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)  
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480) 
cap.set(cv2.CAP_PROP_BUFFERSIZE, 2)  


while True:
    ret, frame = cap.read()
    if not ret:
        print("Error: Could not read frame.")
        break

    cv2.imshow('Dahua Camera', frame)

    # Press 'q' to exit the loop
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

**Reply by pderrenger** [reply]
Hi @Atharva-Malode,

Thanks for trying out the suggestions and sharing your code. It looks like your OpenCV setup is handling the RTSP stream well. If you're experiencing frame loss when integrating with YOLOv8, it might be due to the processing time required for detection, which can delay the frame handling in the loop.

One approach to mitigate this is to separate the frame capture and processing into different threads. This way, one thread can focus on capturing frames from the stream as fast as they come in, while another handles the detection. This can help in not dropping frames due to processing delays.

Here's a basic example of how you might structure this with threading:

```python
import cv2
import threading
from queue import Queue
from ultralytics import YOLO

# Initialize the model
model = YOLO("yolov8n.pt")

# Frame queue
frame_queue = Queue()

def capture_frames():
    cap = cv2.VideoCapture(rtsp_url)
    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break
        if not frame_queue.full():
            frame_queue.put(frame)

    cap.release()

def process_frames():
    while True:
        if not frame_queue.empty():
            frame = frame_queue.get()
            results = model.predict(frame)
            # Process results and display or further processing
            # This is where you'd integrate your existing processing code

# Thread setup
thread_capture = threading.Thread(target=capture_frames)
thread_process = threading.Thread(target=process_frames)

thread_capture.start()
thread_process.start()

thread_capture.join()
thread_process.join()
```

This is just a basic setup and might need adjustments based on your exact requirements. Hope this helps, and let us know how it goes!

**Comment by Atharva-Malode** [comment]
Thank you @pderrenger I was able to solve the problem using Threading where I am processing frames in one thread and capturing frames in another thread also used a queue data structure where I am storing the frames and processing them in first in first out manner. there is a delay of 1 min, but I am not Losing the frames. Anybody facing the same issue can refer to this sample code.

The code is just for reference purpose not to be used in production.

```python
import cv2
import threading
import time
from queue import Queue
from ultralytics.utils.plotting import Annotator
from ultralytics import YOLO


model = YOLO("yolov8n.pt")


frame_queue = Queue()


def capture_frames(rtsp_url):
    cap = cv2.VideoCapture(rtsp_url)
    while True:
        ret, frame = cap.read()
        if not ret:
            break
        if not frame_queue.full():
            frame_queue.put(frame)


def process_frames():
    while True:
        if not frame_queue.empty():
            frame = frame_queue.get()
            
            results = model.predict(frame)

            for r in results:
        
                 annotator = Annotator(frame)
        
                 boxes = r.boxes
                 for box in boxes:
            
                    b = box.xyxy[0]  
                    c = box.cls
                    annotator.box_label(b, model.names[int(c)])
          
                    img = annotator.result()
           
            cv2.imshow('Frame', frame)
            if cv2.waitKey(1) & 0xFF == ord('q'):
                break


rtsp_url = 'rtsp://admin:admin@123@192.168.20.19:554/cam/realmonitor?channel=1&subtype=1'

# Start the frame capture thread
capture_thread = threading.Thread(target=capture_frames, args=(rtsp_url,))
capture_thread.start()

# Start the frame processing thread
process_thread = threading.Thread(target=process_frames)
process_thread.start()

# Wait for the threads to finish
capture_thread.join()
process_thread.join()

**Reply by pderrenger** [reply]
@Atharva-Malode thanks for sharing your solution regarding the frame processing delay when using RTSP streams with YOLOv8! 🚀 Using threading along with a queue to manage the frame flow is a smart approach to ensure that frames are not lost due to I/O blocking in network streams.

For those experiencing similar issues, this example shows how explicitly separating the capturing and processing tasks into different threads can help maintain a smoother frame handling.

Just a reminder for anyone trying similar implementations: make sure to handle thread-safety and clean resource management (like releasing the capture and destroying all OpenCV windows) to avoid any resource leaks or crashes in more extensive applications. Happy coding! 👍
